From 2a310e8f510df74ced4e3f9cc3988b2bafd0af2f Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 21 Oct 2022 12:18:09 +0300
Subject: [PATCH] [SW-107833] gaudi3: update d2d train window

As per recommendation from HW team, in order to get D2D link in low
speed mode as part of rom recovery sequence, the train window default
value needs to be updated (BITSYNC_TRKWIN=0x1F). This patch adds the
support to update train window qualified by fuse bit 25 of row 58.

Change-Id: I0b3fa00ee487c7f76c07e1c1f977b5ab39e14d62
---

diff --git a/arch/gaudi3/arc/dev_efuse.h b/arch/gaudi3/arc/dev_efuse.h
index ac4a9a4..0b1ee3f 100644
--- a/arch/gaudi3/arc/dev_efuse.h
+++ b/arch/gaudi3/arc/dev_efuse.h
@@ -28,6 +28,7 @@
 #define EFUSE_SYSCFG_OOB_PROTOCOL			BIT(19)
 #define EFUSE_SYSCFG_BACKUP_DISABLE			BIT(22)
 #define EFUSE_SYSCFG_D2D_SPI_GW_EN_LOCATION_MASK	BIT(23)
+#define EFUSE_SYSCFG_D2D_UPDATE_TRAIN_WINDOW		BIT(25)
 #define EFUSE_SYSCFG_SKU_MASK				BIT(26)
 #define EFUSE_SYSCFG_PCIE_CONTROL_MASK			BIT(29)
 #define EFUSE_SYSCFG_I3C_EROM_ENABLE_MASK		BIT(31)
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index 7947e14..120190c 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -160,6 +160,23 @@
 		regmap_writel(TIMER_1US, mmD0_DPHY0_CTRL_INIT_PLL_CTRL3 + i * INTER_PHY_DIST);
 	}
 }
+#define BITSYNC_TRKWIN_VAL		0x1F
+static void arcboot_rom_d2d_phy_update_train_window(void)
+{
+	uint32_t val, i;
+
+	if (!(arcboot_rom_global.config_fuse & EFUSE_SYSCFG_D2D_UPDATE_TRAIN_WINDOW))
+		return; /* skip d2d train window update */
+
+	for (i = 0; i < NUM_OF_PHY; i++) {
+		val = regmap_readl(mmD0_DPHY0_CH_BCAST_PHYCTRLREG_RXBITSYNCTMR +
+				i * INTER_PHY_DIST);
+		val |= (BITSYNC_TRKWIN_VAL <<
+			D0_DPHY0_CH_BCAST_PHYCTRLREG_RXBITSYNCTMR_BITSYNC_TRKWIN_S);
+		regmap_writel(val, mmD0_DPHY0_CH_BCAST_PHYCTRLREG_RXBITSYNCTMR +
+				i * INTER_PHY_DIST);
+	}
+}
 
 #define D2D_LINK_STATUS_UDELAY		10000 /* 10 ms */
 static void rom_d2d_link_init(void)
@@ -188,6 +205,8 @@
 	 */
 	arcboot_rom_d2d_phy_wd_hw_errata_fix();
 
+	arcboot_rom_d2d_phy_update_train_window();
+
 	/* Die 0 and Die 1 sync for d2d */
 	if (!arcboot_rom_global.die_id) {
 		/* Indicate slave die that d2d init has started */
